File: java/src/main/java/com/ensta/gestion/Enseignant.java
package com.ensta.gestion;

/**
 * Enseignant : Personne + matière + salaire (TP3).
 * LSP : substituable à Personne, contrat afficherDetails() respecté.
 */
public class Enseignant extends Personne {

    public static final double SALAIRE_MIN = 0.0;

    private String matiere;
    private double salaire;

    public Enseignant(String nom, int age, String matiere, double salaire) {
        super(nom, age);
        setMatiere(matiere);
        setSalaire(salaire);
    }

    public String getMatiere() {
        return matiere;
    }

    public void setMatiere(String matiere) {
        if (matiere == null || matiere.trim().isEmpty()) {
            throw new IllegalArgumentException("La matière ne peut pas être vide.");
        }
        this.matiere = matiere.trim();
    }

    public double getSalaire() {
        return salaire;
    }

    public void setSalaire(double salaire) {
        if (salaire < SALAIRE_MIN) {
            throw new IllegalArgumentException(
                "Le salaire ne peut pas être négatif (reçu : " + salaire + ")."
            );
        }
        this.salaire = salaire;
    }

    @Override
    public String afficherDetails() {
        return "[Enseignant] " + super.afficherDetails()
            + " | Matière : " + matiere
            + " | Salaire : " + String.format("%.2f", salaire) + " €";
    }
}

File: python/enseignant.py
"""Module Enseignant — Hérite de Personne (TP3).

Démontre le polymorphisme : même méthode afficher_details(),
comportement spécifique selon le type réel de l'objet.

LSP : substituable à Personne. Contrat respecté (str retournée, pas d'exception).
"""

from personne import Personne


class Enseignant(Personne):
    """Un enseignant : une Personne avec une matière et un salaire."""

    SALAIRE_MIN = 0.0

    def __init__(self, nom: str, age: int, matiere: str, salaire: float):
        super().__init__(nom, age)
        if not matiere or not matiere.strip():
            raise ValueError("La matière ne peut pas être vide.")
        self.__matiere = matiere.strip()
        self.salaire = salaire  # passage par setter → validation

    # ----- matiere -----
    @property
    def matiere(self) -> str:
        return self.__matiere

    @matiere.setter
    def matiere(self, valeur: str) -> None:
        if not valeur or not valeur.strip():
            raise ValueError("La matière ne peut pas être vide.")
        self.__matiere = valeur.strip()

    # ----- salaire -----
    @property
    def salaire(self) -> float:
        return self.__salaire

    @salaire.setter
    def salaire(self, valeur: float) -> None:
        if not isinstance(valeur, (int, float)) or isinstance(valeur, bool):
            raise ValueError("Le salaire doit être un nombre.")
        if valeur < self.SALAIRE_MIN:
            raise ValueError(
                f"Le salaire ne peut pas être négatif (reçu : {valeur})."
            )
        self.__salaire = float(valeur)

    # ----- polymorphisme TP3 -----
    def afficher_details(self) -> str:
        base = super().afficher_details()
        return (
            f"[Enseignant] {base} | Matière : {self.__matiere} | "
            f"Salaire : {self.__salaire:.2f} €"
        )

    def __repr__(self) -> str:
        return (
            f"Enseignant(nom={self.nom!r}, age={self.age}, "
            f"matiere={self.__matiere!r}, salaire={self.__salaire})"
        )
